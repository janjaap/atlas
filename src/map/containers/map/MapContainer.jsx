import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { LatLngBounds } from 'leaflet';

import { toggleMapFullscreen } from '../../../shared/ducks/ui/ui';

import DrawTool from '../../containers/draw-tool/DrawToolContainer'; //eslint-disable-line
import ToggleFullscreen from '../../components/toggle-fullscreen/ToggleFullscreen';

import LeafletContainer from '../leaflet/LeafletContainer';
import MapPanelContainer from '../../containers/panel/MapPanelContainer';
import MapPreviewPanelContainer from '../../containers/preview-panel/MapPreviewPanelContainer';
import MapEmbedButton from '../../components/map-embed-button/MapEmbedButton';

import getEmbedLink from '../../ducks/embed/embed';
import piwikTracker from '../../../shared/services/piwik-tracker/piwik-tracker';

export const overrideLeafletGetBounds = (map) => {
  // We override here the getBounds method of Leaflet
  // To ensure the full coverage of the visible area
  // of the NonTiledLayer layer types
  map.getBounds = () => { // eslint-disable-line no-param-reassign
    const bounds = map.getPixelBounds();
    const sw = map.unproject(bounds.getBottomLeft());
    const ne = map.unproject(bounds.getTopRight());

    const latLngBounds = (new LatLngBounds(sw, ne)).pad(0.02);
    return latLngBounds;
  };
};

const mapStateToProps = (state) => ({
  isFullscreen: state.ui.isMapFullscreen,
  drawMode: state.map.drawingMode,
  embedLink: getEmbedLink(state)
});

const mapDispatchToProps = (dispatch) => bindActionCreators({
  onToggleFullscreen: toggleMapFullscreen
}, dispatch);

class MapContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      leafletInstance: null
    };
    this.setLeafletInstance = this.setLeafletInstance.bind(this);
    this.onToggleFullscreen = this.onToggleFullscreen.bind(this);
  }

  // TODO DP-6031: Create Redux Middelware, map Piwik events to ACTIONS
  onToggleFullscreen() {
    const { isFullscreen, onToggleFullscreen } = this.props;
    const piwik = {
      TRACK_EVENT: 'trackEvent',
      MINIMIZE_MAP: 'map-minimize',
      NAVIGATION: 'navigation'
    };

    if (isFullscreen) {
      piwikTracker([piwik.TRACK_EVENT, piwik.NAVIGATION,
        piwik.MINIMIZE_MAP, window.document.title]);
    }

    return onToggleFullscreen();
  }

  setLeafletInstance(leafletInstance) {
    overrideLeafletGetBounds(leafletInstance);
    this.setState({ leafletInstance });
  }

  render() {
    return (
      <div className={`c-map c-map--drawing-mode-${this.props.drawMode} qa-map-container`}>
        <LeafletContainer
          getLeafletInstance={this.setLeafletInstance}
        />
        {
          this.state.leafletInstance && (
            <DrawTool
              leafletInstance={this.state.leafletInstance}
            />
          )
        }
        <ToggleFullscreen
          isFullscreen={this.props.isFullscreen}
          onToggleFullscreen={this.onToggleFullscreen}
        />
        <MapPanelContainer />
        {
          this.props.embedLink.length ? (
            <MapEmbedButton link={this.props.embedLink} />
          ) : ''
        }
        <MapPreviewPanelContainer />
      </div>
    );
  }
}

MapContainer.contextTypes = {
  store: PropTypes.object.isRequired
};

MapContainer.defaultProps = {
  geometry: null,
  leafletInstance: null,
  drawMode: 'none',
  embedLink: ''
};

MapContainer.propTypes = {
  isFullscreen: PropTypes.bool.isRequired,
  onToggleFullscreen: PropTypes.func.isRequired,
  drawMode: PropTypes.string,
  embedLink: PropTypes.string
};

export default connect(mapStateToProps, mapDispatchToProps)(MapContainer);
