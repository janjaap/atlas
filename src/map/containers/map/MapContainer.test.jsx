import React from 'react';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';

import MapContainer, { overrideLeafletGetBounds } from './MapContainer';

import piwikTracker from '../../../shared/services/piwik-tracker/piwik-tracker';

jest.mock('../../../shared/services/piwik-tracker/piwik-tracker');

let initialState;

describe('overrideLeafletGetBounds', () => {
  it('should overide the getBounds', () => {
    const bl = { x: 482988, y: -551411 };
    const tr = { x: 484747, y: -552666 };
    const pixelBounds = {
      max: { x: 484747, y: -551411 },
      min: { x: 482988, y: -552666 },
      getBottomLeft: () => bl,
      getTopRight: () => tr
    };

    const sw = { lat: 52.35884827870951, lng: 4.877935872162896 };
    const ne = { lat: 52.36841402615222, lng: 4.8995211251421305 };

    const resultBounds = {
      _northEast: {
        lat: 52.36860534110108,
        lng: 4.899952830201715
      },

      _southWest: {
        lat: 52.35865696376065,
        lng: 4.877504167103311
      }
    };
    const map = {
      getPixelBounds: jest.fn(() => pixelBounds),
      unproject: jest.fn((p) => {
        if (p.x === 482988) return sw;
        return ne;
      })
    };

    overrideLeafletGetBounds(map);
    expect(map.getBounds).toBeDefined();
    const bounds = map.getBounds();
    expect(bounds).toEqual(resultBounds);
  });
});

describe('MapContainer', () => {
  beforeEach(() => {
    initialState = {
      ui: {
        isMapFullscreen: false
      },
      map: {
        drawingMode: 'none'
      }
    };
  });

  it('should render', () => {
    const store = configureMockStore()({ ...initialState });
    const wrapper = shallow(<MapContainer />, { context: { store } }).dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('should render with drawingmode: draw', () => {
    const store = configureMockStore()({ ...initialState, map: { drawingMode: 'draw' } });
    const wrapper = shallow(<MapContainer />, { context: { store } }).dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('should set the leaflet instance state', () => {
    const store = configureMockStore()({ ...initialState });
    const wrapper = shallow(<MapContainer />, { context: { store } }).dive();
    wrapper.instance().setLeafletInstance({});
    expect(wrapper.instance().state.leafletInstance).toBeTruthy();
    expect(wrapper.instance().state.leafletInstance.getBounds).toBeDefined();
  });

  it('should hide the fullscreen and send piwik action', () => {
    const store = configureMockStore()({ ...initialState, ui: { isMapFullscreen: true } });
    const wrapper = shallow(<MapContainer />, { context: { store } }).dive();
    wrapper.instance().onToggleFullscreen();
    piwikTracker.mockImplementation(() => jest.fn());
    expect(piwikTracker).toHaveBeenCalled();
    // should not send another piwik event if map is maximized
    wrapper.setProps({ isFullscreen: false });
    wrapper.instance().onToggleFullscreen();
    expect(piwikTracker).toHaveBeenCalledTimes(1);
  });
});
