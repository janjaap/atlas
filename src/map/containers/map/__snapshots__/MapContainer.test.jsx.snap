// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapContainer should render 1`] = `
<div
  className="c-map c-map--drawing-mode-none qa-map-container"
>
  <Connect(LeafletContainer)
    getLeafletInstance={[Function]}
  />
  <ToggleFullscreen
    isFullscreen={false}
    onToggleFullscreen={[Function]}
  />
  <Connect(MapPanelContainer) />
  <Connect(MapPreviewPanel) />
</div>
`;

exports[`MapContainer should render with drawingmode: draw 1`] = `
<div
  className="c-map c-map--drawing-mode-draw qa-map-container"
>
  <Connect(LeafletContainer)
    getLeafletInstance={[Function]}
  />
  <ToggleFullscreen
    isFullscreen={false}
    onToggleFullscreen={[Function]}
  />
  <Connect(MapPanelContainer) />
  <Connect(MapPreviewPanel) />
</div>
`;
